model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt

```diff
@@ -337,3 +371,56 @@ class RepositoriesManager(val client: LocalModelClient) {
 }
 
 class RepositoryAlreadyExistsException(val name: String) : IllegalStateException("Repository '$name' already exists")
+
+sealed interface ObjectData {
+    suspend fun asMap(): Map<String, String>
+    fun asFlow(): Flow<Pair<String, String>>
+
+    companion object {
+        val empty = ObjectDataMap(emptyMap())
+    }
+}
+
+class ObjectDataMap(private val byHashObjects: Map<String, String>) : ObjectData {
+    init {
+        HashUtil.checkObjectHashes(byHashObjects)
+    }
+
+    override suspend fun asMap(): Map<String, String> = byHashObjects
+    override fun asFlow(): Flow<Pair<String, String>> = byHashObjects.entries.asFlow().map { it.toPair() }
+}
+
+class ObjectDataFlow(private val hashObjectFlow: Flow<Pair<String, String>>) : ObjectData {
+    override suspend fun asMap(): Map<String, String> = hashObjectFlow.toMap()
+    override fun asFlow(): Flow<Pair<String, String>> = hashObjectFlow
+}
+
+private fun Flow<Pair<String, String>>.checkObjectHashes(): Flow<Pair<String, String>> {
+    return onEach { HashUtil.checkObjectHash(it.first, it.second) }
+}
+
+class VersionDeltaCache(val store: IDeserializingKeyValueStore) {
+    private val cacheMap = LRUMap<Pair<String, String?>, SoftReference<Deferred<ObjectDataMap>>>(10)
+    private val cacheMapMutex = Mutex()
+
+    suspend fun getOrComputeDelta(versionHash: String, baseVersionHash: String): ObjectDataMap {
+        val deferredDelta = cacheMapMutex.withLock {
```

Comment:
The mutex is unnecessary. `synchronized (cacheMap)` can be used, because the actual computation happens afterwards inside the `async` coroutine.

Reply:
Done.